com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.refactor.cs

```diff
@@ -0,0 +1,262 @@
+    {
+        /// <summary>
+        /// For each scene:
+        /// - Make names of timeline object unique - for unique referencing later on
+        /// - Copy prefab instances (and upgrade the copy), build conversion link and collect timeline references
+        /// </summary>
+        /// <param name="conversionLinksPerScene">Key: scene index, Value: List of conversion links</param>
+        /// <param name="renameMap">Timeline rename mapping</param>
+        void PrepareUpgrades(out Dictionary<int, List<ConversionLink>> conversionLinksPerScene, out Dictionary<string, string> renameMap)
+        {
+            conversionLinksPerScene = new Dictionary<int, List<ConversionLink>>();
+            renameMap = new Dictionary<string, string>();
+            for (var s = 0; s < m_SceneManager.SceneCount; ++s)
+            {
+                var scene = OpenScene(s);
+
+                // Make timeline names unique
+#if CINEMACHINE_TIMELINE
+                var directors = TimelineManager.GetPlayableDirectors(scene);
+                foreach (var director in directors)
+                {
+                    var originalName = director.name;
+                    director.name = GUID.Generate().ToString();
+                    renameMap.Add(director.name, originalName); // key = guid, value = originalName
+                }
+#endif
+
+                // CopyPrefabInstances, give unique names, create conversion links, collect timeline references
+                // Upgrade prefab instance copies of referencables only
+                var conversionLinks = new List<ConversionLink>();
+                var timelineManager = new TimelineManager(scene);
+                var allPrefabInstances = new List<GameObject>();
+
+                for (var p = 0; p < m_PrefabManager.PrefabCount; ++p)
+                {
+                    allPrefabInstances.AddRange(
+                        PrefabManager.FindAllInstancesOfPrefabEvenInNestedPrefabs(scene,
+                            m_PrefabManager.GetPrefabAssetPath(p)));
+                }
+
+                var upgradedObjects = new HashSet<GameObject>();
+                var upgradable =
+                    GetUpgradables(allPrefabInstances.ToArray(), m_ObjectUpgrader.RootUpgradeComponentTypes, true);
+                foreach (var go in upgradable)
+                {
+                    if (upgradedObjects.Contains(go))
+                        continue; // Ignore if already converted (this can happen in nested prefabs)
+                    upgradedObjects.Add(go);
+#if CINEMACHINE_TIMELINE
+                    var originalVcam = go.GetComponent<CinemachineVirtualCameraBase>();
+                    var timelineReferences = timelineManager.GetTimelineReferences(originalVcam);
+#endif
+                    var convertedCopy = Object.Instantiate(go);
+                    UpgradeObjectComponents(convertedCopy, null);
+                    m_ObjectUpgrader.DeleteObsoleteComponents(convertedCopy);
```

Comment:
Is there a reason to delete these now?  It will destroy any references to the deleted components.  Best to delete at the end of the process, I think.